### PR TITLE
Add bitwise and compare ops

### PR DESCRIPTION
**Pull request by xytintel**
e.g. isnan, clamp, clamp_min, clamp_max, bitwise_and, bitwise_or, bitwise_xor, bitwise_not, _local_scalar_dense

### CODE REVIEW

File: src/aten/XPUScalar.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            c10::nullopt, /* device */
+            true, /* pin_memory */
+            c10::nullopt /* memory format */
+        );
+        auto q = at::xpu::getCurrentSYCLQueue();
+
+        q.memcpy(
+            (void*)value.const_data_ptr<scalar_t>(),
+            self.const_data_ptr<scalar_t>(),
+            sizeof(scalar_t));
```
Comment (fengyuan14) [inline src/aten/XPUScalar.cpp:30]: Need synchronize the copy, or the value on host could not be valid.

File: src/aten/CompareOps.cpp [hunk 2]
```diff
@@ -248,4 +250,117 @@ Tensor& XPUNativeFunctions::ge_out(
...
+Tensor& XPUNativeFunctions::isnan_out(const Tensor& self, Tensor& out) {
+  return XPUNativeFunctions::ne_out(self, self, out);
+}
+
+Tensor& XPUNativeFunctions::clamp_out(
+    const Tensor& self,
+    const c10::optional<Scalar>& min,
+    const c10::optional<Scalar>& max,
+    Tensor& out) {
+  if (min && max) {
```
Comment (fengyuan14) [inline src/aten/CompareOps.cpp:266]: Please align PyTorch implementation.
image

File: src/aten/CompareOps.cpp [hunk 3]
```diff
@@ -248,4 +250,117 @@ Tensor& XPUNativeFunctions::ge_out(
...
+    XPUNativeFunctions::clamp_min_out(self, *min, out);
+  }
+  return out;
+}
+
+Tensor& XPUNativeFunctions::clamp_min_out(
+    const Tensor& self,
+    const Scalar& min,
+    Tensor& out) {
+  auto iter = TensorIterator::unary_op(out, self);
```
Comment (fengyuan14) [inline src/aten/CompareOps.cpp:281]: Please align PyTorch implementation.
image

File: src/aten/CompareOps.cpp [hunk 4]
```diff
@@ -248,4 +250,117 @@ Tensor& XPUNativeFunctions::ge_out(
...
+  native::xpu::clamp_min_kernel(iter, min);
+  return out;
+}
+
+Tensor& XPUNativeFunctions::clamp_max_out(
+    const Tensor& self,
+    const Scalar& max,
+    Tensor& out) {
+  auto iter = TensorIterator::unary_op(out, self);
```
Comment (fengyuan14) [inline src/aten/CompareOps.cpp:290]: Please align with PyTorch implementation
image

File: src/aten/sycl/CompareKernels.cpp [hunk 5]
```diff
@@ -50,6 +51,60 @@ struct GeFunctor {
   }
 };
 
+template <typename opmath_t>
+struct ClampFunctor {
+  opmath_t operator()(opmath_t v) const {
```
Comment (fengyuan14) [inline src/aten/sycl/CompareKernels.cpp:56]: Lack of sanity check of `lower_` and `upper_`. Please align with PyTorch implementation.
image

File: src/aten/sycl/CompareKernels.cpp [hunk 6]
```diff
@@ -50,6 +51,60 @@ struct GeFunctor {
...
+
+ private:
+  opmath_t lower_;
+  opmath_t upper_;
+};
+
+template <typename opmath_t>
+struct ClampMinFunctor {
+  opmath_t operator()(opmath_t v) const {
+    if (_isnan(v)) {
```
Comment (fengyuan14) [inline src/aten/sycl/CompareKernels.cpp:73]: Please align wth PyTorch implementation.
image

File: src/aten/sycl/CompareKernels.cpp [hunk 7]
```diff
@@ -50,6 +51,60 @@ struct GeFunctor {
...
+  }
+  ClampMinFunctor(opmath_t lower) : lower_(lower) {}
+
+ private:
+  opmath_t lower_;
+};
+
+template <typename opmath_t>
+struct ClampMaxFunctor {
+  opmath_t operator()(opmath_t v) const {
```
Comment (fengyuan14) [inline src/aten/sycl/CompareKernels.cpp:87]: ditto

File: src/aten/sycl/CompareKernels.cpp [hunk 8]
```diff
@@ -116,6 +171,41 @@ void ge_kernel(TensorIteratorBase& iter) {
...
 
+void clamp_kernel(
+    TensorIteratorBase& iter,
+    const Scalar& min_value,
+    const Scalar& max_value) {
+  AT_DISPATCH_ALL_TYPES_AND2(
+      kHalf, kBFloat16, iter.dtype(), "clamp_xpu", [&]() {
+        auto lower = min_value.to<scalar_t>();
+        auto upper = max_value.to<scalar_t>();
+        gpu_kernel(iter, ClampFunctor<scalar_t>(lower, upper));
```
Comment (fengyuan14) [inline src/aten/sycl/CompareKernels.cpp:182]: Cannot find opmath definition inside of functor? Missing?

File: src/aten/sycl/CompareKernels.cpp [hunk 9]
```diff
@@ -116,6 +171,41 @@ void ge_kernel(TensorIteratorBase& iter) {
...
+      });
+}
+
+void clamp_min_kernel(TensorIteratorBase& iter, const Scalar& min_value) {
+  AT_DISPATCH_ALL_TYPES_AND2(
+      kHalf, kBFloat16, iter.dtype(), "clamp_min_xpu", [&]() {
+        auto lower = min_value.to<scalar_t>();
+        gpu_kernel(iter, ClampMinFunctor<scalar_t>(lower));
```
Comment (fengyuan14) [inline src/aten/sycl/CompareKernels.cpp:190]: ditto, checking opmath.

File: src/aten/sycl/CompareKernels.cpp [hunk 10]
```diff
@@ -116,6 +171,41 @@ void ge_kernel(TensorIteratorBase& iter) {
...
+      });
+}
+
+void clamp_max_kernel(TensorIteratorBase& iter, const Scalar& max_value) {
+  AT_DISPATCH_ALL_TYPES_AND2(
+      kHalf, kBFloat16, iter.dtype(), "clamp_max_xpu", [&]() {
+        auto upper = max_value.to<scalar_t>();
+        gpu_kernel(iter, ClampMaxFunctor<scalar_t>(upper));
```
Comment (fengyuan14) [inline src/aten/sycl/CompareKernels.cpp:198]: ditto, checking opmath.

File: src/aten/sycl/CompareKernels.cpp (hunk 8 above)
Comment (fengyuan14) [reply src/aten/sycl/CompareKernels.cpp:182]: No need opmath. Dismiss.

**Comment by xytintel** [comment]
split pr

**Pull request closed without merging**